components/script/dom/globalscope.rs

```diff
@@ -207,6 +351,347 @@ impl GlobalScope {
+        };
+        // Let entangled port know it was shipped as well.
+        if let Some(port) = self.message_ports.borrow_mut().get(&entangled_id) {
+            port.set_has_been_shipped();
+        }
+    }
+
+    /// Enable a port.
+    pub fn start_message_port(&self, port_id: &MessagePortId) {
+        if let Some(port) = self.message_ports.borrow_mut().get(port_id) {
+            return port.start(self);
+        }
+        unreachable!("Port whose start was called should exist");
+    }
+
+    /// Close a port.
+    pub fn close_message_port(&self, port_id: &MessagePortId) {
+        if let Some(port) = self.message_ports.borrow_mut().get(port_id) {
+            return port.close();
+        }
+        unreachable!("Port whose close was called should exist");
+    }
+
+    /// Get the onmessage handler for a port.
+    pub fn get_message_port_onmessage(
+        &self,
+        port_id: &MessagePortId,
+    ) -> Option<Rc<EventHandlerNonNull>> {
+        if let Some(listener) = self.message_listeners.borrow().get(port_id) {
+            return listener.clone();
+        }
+        None
+    }
+
+    /// Set the onmessage handler for a port.
+    pub fn set_message_port_onmessage(
+        &self,
+        port_id: &MessagePortId,
+        listener: Option<Rc<EventHandlerNonNull>>,
+    ) {
+        self.message_listeners
+            .borrow_mut()
+            .insert(port_id.clone(), listener);
+        if let Some(port) = self.message_ports.borrow_mut().get(port_id) {
+            return port.start(self);
+        }
+        unreachable!("Port whose onmessage was set should exist");
+    }
+
+    /// Post a message via a port.
+    pub fn post_messageport_msg(&self, port_id: MessagePortId, task: PortMessageTask) {
+        if let Some(port) = self.message_ports.borrow_mut().get_mut(&port_id) {
+            return port.post_message(self, task);
+        }
+        unreachable!("Port whose postMessage was called should exist");
+    }
+
+    /// Route the task to be handled by the relevant port.
+    pub fn route_task_to_port(&self, port_id: MessagePortId, task: PortMessageTask) {
+        if self
+            .message_ports_transferred
+            .borrow_mut()
+            .contains(&port_id)
+        {
+            // If the port has already been transferred before this message arrived, re-route it.
+            // This should only happen initially, when the entangled ports sends a message
+            // and hasn't been updated about the transfer yet.
+            let _ = self
+                .script_to_constellation_chan()
+                .send(ScriptMsg::RerouteMessagePort(port_id.clone(), task));
+            return;
+        }
+
+        // If the port is not enabled yet, or if it is awaiting the completion of it's transfer,
+        // the task will be buffered and dispatched upon enablement or completion of the transfer.
+        let should_dispatch = if let Some(port) = self.message_ports.borrow_mut().get_mut(&port_id)
```

Comment:
for this one, we can use [`map_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) to give a default value for the `None` case 👀